runtime/runc_jailer.go

```diff
@@ -353,36 +367,68 @@ func (j runcJailer) jailerCommand(containerName string, isDebug bool) *exec.Cmd
 }
 
 // overwriteConfig will set the proper default values if a field had not been set.
-//
-// TODO: Add netns
-func (j runcJailer) overwriteConfig(cfg *Config, socketPath, configPath string) error {
-	spec := specs.Spec{}
-	configBytes, err := ioutil.ReadFile(configPath)
-	if err != nil {
-		return err
-	}
+func (j *runcJailer) overwriteConfig(cfg *Config, machineConfig *firecracker.Config, socketPath, configPath string) error {
+	var err error
+	j.once.Do(func() {
+		if configSpec == nil {
+			spec := specs.Spec{}
+			var configBytes []byte
+			configBytes, err = ioutil.ReadFile(configPath)
+			if err != nil {
+				return
+			}
 
-	if err := json.Unmarshal(configBytes, &spec); err != nil {
-		return err
-	}
+			if err = json.Unmarshal(configBytes, &spec); err != nil {
+				return
+			}
 
-	if spec.Process.User.UID != 0 ||
-		spec.Process.User.GID != 0 {
-		return fmt.Errorf(
-			"using UID %d and GID %d, these values must not be set",
-			spec.Process.User.UID,
-			spec.Process.User.GID,
-		)
-	}
+			configSpec = &spec
+
+			if spec.Process.User.UID != 0 ||
+				spec.Process.User.GID != 0 {
+				err = fmt.Errorf(
+					"using UID %d and GID %d, these values must not be set",
+					spec.Process.User.UID,
+					spec.Process.User.GID,
+				)
+				return
+			}
 
-	spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+			spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+			spec.Root.Path = rootfsFolder
+			spec.Root.Readonly = false
+		}
+	})
 
-	spec.Root.Path = rootfsFolder
-	spec.Root.Readonly = false
+	if err != nil {
+		return err
+	}
+
+	spec := *configSpec
 	spec.Process.User.UID = j.uid
 	spec.Process.User.GID = j.gid
 
-	configBytes, err = json.Marshal(&spec)
+	// remove the network namespace if there exists a CNI
+	if hasCNI(machineConfig.NetworkInterfaces) {
```

Comment:
Another thought on all this, `overwriteConfig` runs as part of a handler, which means that it runs after [`NewMachine`](https://github.com/firecracker-microvm/firecracker-containerd/blob/5f3e0c74a7c707ac7e0cd0a8f1466abff75d47de/runtime/service.go#L496) is called, right?

If so, `NewMachine` is where the SDK [sets the network namespace if there is a CNI interface](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/eec83f3dcb7b4eb4eeb1e2c5f975715a25f79e96/machine.go#L304-L306). So then because it's in a handler (with signature `func(ctx context.Context, m *firecracker.Machine)`), we can access that netns value at `m.Cfg.NetNS`, right? If so, then `overwriteConfig` shouldn't actually need to run `hasCNI`, it should be able to just read `m.Cfg.NetNS` and fill in the runc config if a netns is specified there.

I still think the better long-term cleanup of this is probably to move all this jailing to the go-sdk, but if the above works then it should simplify a little bit in the short term at least.

Reply:
Fixed!

Yea, I agree a better solution would be to have a jailer interface in the SDK. However, the logic would probably still live here, but would instead just satisfy the interface and pass that into the SDK's config.